File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        int length = length(head);
        if(length == 0){
            return null;
        }
        if(n == 0 || n > length){
            return head;
        }
        if(length == 1 && n == 1){
            return null;
        }
        if(n == length){
            head = head.next;
            return head;
        }
        ListNode iter = head;
        for(int i =1; i < length-n; i++){
            iter = iter.next;
        }
        if(n == 1){
            iter.next = null;
            return head;
        }
        iter.next = iter.next.next;
        return head;
    }
    public int length(ListNode head){
        ListNode iter = head;
        int count = 0;
        while(iter != null){
            iter = iter.next;
            count++;
        }
        return count;
    }
}

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* removeNthFromEnd(struct ListNode* head, int n) {
    struct ListNode* dummy = (struct ListNode*)malloc(sizeof(struct ListNode));
    dummy->next = head;
    struct ListNode* current = dummy;
    int length = 0;
    while(current != NULL){
        length++;
        current = current->next;
    }
    current = dummy;
    length -= n;
    for(int i=1; i<length; i++){
        current = current->next;
    }
    current->next = current->next->next;
    return dummy->next;
}